pkg/json/parse.go

```diff
@@ -111,19 +112,62 @@ func (ing *Parser) parseDoc(ctx parserContext, input jsonom.Node) (*ParsedDocNod
 	return ing.parseValue(ctx, input.(*jsonom.Value))
 }
 
+var PolyHintBlock = func(*jsonom.KeyValue) {}
+
 func (ing *Parser) parseObject(ctx parserContext, input *jsonom.Object) (*ParsedDocNode, error) {
 	// An object node
 	if ctx.schemaNode != nil {
 		if !ctx.schemaNode.HasLabel(ls.AttributeTypeObject) {
 			return nil, ls.ErrSchemaValidation{Msg: fmt.Sprintf("An object is expected here but found %s", ctx.schemaNode.GetLabels()), Path: ctx.path.Copy()}
 		}
 	}
+	if ing.discriminator == nil {
+		ing.discriminator = make(map[*lpg.Node][]*lpg.Node)
+	}
+
+	parseNewCtx := func(ctx parserContext, node *lpg.Node, keyvalue func(input *jsonom.Object) *jsonom.KeyValue) (*ParsedDocNode, error) {
+		newCtx := ctx
+		newCtx.schemaNode = node
+		newCtx.path = newCtx.path.AppendString(keyvalue(input).Key())
+		parsed, err := ing.parseDoc(newCtx, keyvalue(input).Value())
+		if err != nil {
+			return nil, err
+		}
+		return parsed, err
+	}
+
+	if discrims, cached := ing.discriminator[ctx.schemaNode]; cached {
+		for _, snode := range discrims {
+			_, err := parseNewCtx(ctx, snode, func(input *jsonom.Object) *jsonom.KeyValue {
+				return input.Get(ls.AsPropertyValue(snode.GetProperty(ls.AttributeNameTerm)).AsString())
+			})
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
 	// There is a schema node for this node. It must be an object
 	nextNodes, err := ing.getObjectNodes(ctx.schemaNode)
 	if err != nil {
 		return nil, err
 	}
 
+	for _, sln := range nextNodes {
```

Comment:
To fill up the cache:
if ing.discriminator[ctx.schemaNode] does not exist:
  iterate nextNodes
  if schemanode has discriminator, add it to the ing.discriminator cache